File: DataStructure/Queue/src/Queue.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node* link;
} Node;

typedef struct {
    Node* front;
    Node* rear;
} Queue;

void init(Queue* q) {
    q->front = q->rear = NULL;
}

void enqueue(Queue* q, int data) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->data = data;
    node->link = NULL;
    if (q->rear) q->rear->link = node;
    else q->front = node;
    q->rear = node;
}

int dequeue(Queue* q) {
    if (!q->front) return -1;
    Node* tmp = q->front;
    int data = tmp->data;
    q->front = q->front->link;
    if (!q->front) q->rear = NULL;
    free(tmp);
    return data;
}

File: DataStructure/Queue/src/Queue.java
class Node {
    int data;
    Node link;

    Node(int data) {
        this.data = data;
        this.link = null;
    }
}

class Queue {
    private Node front, rear;

    public Queue() {
        front = rear = null;
    }

    public void enqueue(int item) {
        Node node = new Node(item);
        if (rear != null) rear.link = node;
        else front = node;
        rear = node;
    }

    public int dequeue() {
        if (front == null) throw new RuntimeException("Empty Queue");
        int data = front.data;
        front = front.link;
        if (front == null) rear = null;
        return data;
    }
}